### PR TITLE
knex: Add JoinClause 'on' variations

### PR DESCRIPTION
**Pull request by HillTravis**
Added `onIn`, `notOnIn`, `onNull`, `onNotNull`, `onExists`, `onNotExists`, `onBetween`, and `onNotBetween` to `JoinClause` interface.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
  - Simply added pieces that were missing from the main documentation, no specific URL needed.
- [ ] Increase the version number in the header if appropriate.
  - No version number in URL.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.
  - Changes are not substantial, followed existing conventions.

### CODE REVIEW

**Comment by tkqubo** [comment]
👍 

But

```ts
onBetween(column1: string, range: any[])
```
could be

```ts
onBetween(column1: string, range: [any, any])
```

to avoid an array with only one element passed to the 2nd argument, which would fail in assertion here:

https://github.com/tgriesser/knex/blob/67b62d3213aa6ade68bdf84cb95b3410a1fd8053/src/query/joinclause.js#L72-L74

**Comment by HillTravis** [comment]
@tkqubo Good call. Updated.

File: types/knex/index.d.ts [hunk 1]
```diff
@@ -206,6 +206,14 @@ declare namespace Knex {
         orOn(column1: string, column2: string): JoinClause;
         orOn(column1: string, raw: Raw): JoinClause;
         orOn(column1: string, operator: string, column2: string): JoinClause;
+        onIn(column1: string, values: any[]): JoinClause;
+        onNotIn(column1: string, values: any[]): JoinClause;
+        onNull(column1: string): JoinClause;
+        onNotNull(column1: string): JoinClause;
+        onExists(callback: Function): JoinClause;
```
Comment (andy-ms) [inline types/knex/index.d.ts]: Could this have a better type than `Function`, such as (just guessing) `(result: boolean) => void`?

**Comment by HillTravis** [comment]
@andy-ms I changed it to `() => void`. The reason I hesitated to do that at first was that the documentation isn't always completely clear about all of the possibilities of that function signature, so I was erring on the side of caution. Also, that was how it was mostly done throughout the rest of the existing file.

@tkqubo I also added the variations for `and` and `or` for each of those `on` variations that I added, and added tests for them. Not sure if you need to re-approve for any reason.

**Pull request merged**